src/ILCompiler.Compiler/src/Compiler/Compilation.cs

```diff
@@ -431,6 +437,64 @@ public void RootReadOnlyDataBlob(byte[] data, int alignment, string reason, stri
                 _graph.AddRoot(_factory.ReadOnlyDataBlob(exportName, data, alignment), reason);
             }
         }
+
+        private sealed class ILCache : LockFreeReaderHashtable<MethodDesc, ILCache.MethodILData>
+        {
+            public ILProvider ILProvider { get; }
+
+            public ILCache(ILProvider provider)
+            {
+                ILProvider = provider;
+            }
+
+            protected override int GetKeyHashCode(MethodDesc key)
+            {
+                return key.GetHashCode();
+            }
+            protected override int GetValueHashCode(MethodILData value)
+            {
+                return value.Method.GetHashCode();
+            }
+            protected override bool CompareKeyToValue(MethodDesc key, MethodILData value)
+            {
+                return Object.ReferenceEquals(key, value.Method);
+            }
+            protected override bool CompareValueToValue(MethodILData value1, MethodILData value2)
+            {
+                return Object.ReferenceEquals(value1.Method, value2.Method);
+            }
+            protected override MethodILData CreateValueFromKey(MethodDesc key)
+            {
+                return new MethodILData() { Method = key, MethodIL = ILProvider.GetMethodIL(key) };
+            }
+
+            internal class MethodILData
+            {
+                public MethodDesc Method;
+                public MethodIL MethodIL;
+            }
+        }
+
+        private sealed class CombinedILProvider : ILProvider
+        {
+            private readonly ILProvider _primaryILProvider;
+            private readonly PInvokeILProvider _pinvokeProvider;
+
+            public CombinedILProvider(ILProvider primaryILProvider, PInvokeILProvider pinvokeILProvider)
+            {
+                _primaryILProvider = primaryILProvider;
+                _pinvokeProvider = pinvokeILProvider;
+            }
+
+            public override MethodIL GetMethodIL(MethodDesc method)
+            {
+                MethodIL result = _primaryILProvider.GetMethodIL(method);
```

Comment:
`method.IsPInvoke ? _pinvokeProvider.GetMethodIL(method) :  _primaryILProvider.GetMethodIL(method)` ?

Reply:
Looks more readable - my micro-optimization is not worth it, fixed.

Reply:
I had to roll that back. This caused trouble with calli marshalling method thunks - they have `IsPInvoke == true`, but `PInvokeILProvider` can't provide method body for them because they're `ILStubMethod` and they provide their own IL.